File: src/state/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { Category } from "../types";

export interface TickEvent {
  date_key: string;
  minute_of_day: number;
  category: Category;
  source_key: string | null;
  source_title: string | null;
}

export interface CurrentActivityEvent {
  process: string | null;
  title: string | null;
  category: Category;
  source_key: string;
  idle_ms: number;
  paused: boolean;
  idle_break: boolean;
  afk_threshold_ms: number;
}

export function onTick(handler: (e: TickEvent) => void): Promise<UnlistenFn> {
  return listen<TickEvent>("hm:tick", (event) => handler(event.payload));
}

export function onCurrentActivity(
  handler: (e: CurrentActivityEvent) => void,
): Promise<UnlistenFn> {
  return listen<CurrentActivityEvent>("hm:current-activity", (event) => handler(event.payload));
}

/** Fires when stored minutes change out-of-band (reclassify, backfill). */
export function onDayChanged(handler: () => void): Promise<UnlistenFn> {
  return listen("hm:day-changed", () => handler());
}

File: src-tauri/src/events.rs
use serde::Serialize;
use tauri::Emitter;

#[derive(Debug, Clone, Serialize)]
pub struct TickPayload {
    pub date_key: String,
    pub minute_of_day: u16,
    pub category: &'static str,
    pub source_key: Option<String>,
    pub source_title: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CurrentActivityPayload {
    pub process: Option<String>,
    pub title: Option<String>,
    pub category: &'static str,
    pub source_key: String,
    pub idle_ms: u32,
    pub paused: bool,
    /// True when idle has crossed the threshold and this minute is being counted
    /// as a break (its own entity) rather than the foreground app.
    pub idle_break: bool,
    /// Current idle→break threshold in ms (so the UI can show it for debugging).
    pub afk_threshold_ms: u32,
}

pub fn emit_tick<R: tauri::Runtime>(app: &tauri::AppHandle<R>, payload: TickPayload) {
    if let Err(err) = app.emit("hm:tick", payload) {
        tracing::warn!(?err, "emit hm:tick failed");
    }
}

pub fn emit_current_activity<R: tauri::Runtime>(
    app: &tauri::AppHandle<R>,
    payload: CurrentActivityPayload,
) {
    if let Err(err) = app.emit("hm:current-activity", payload) {
        tracing::warn!(?err, "emit hm:current-activity failed");
    }
}

/// Signals that stored minutes changed out-of-band (reclassify, backfill) so
/// open windows re-fetch the day instead of waiting for the next tick.
pub fn emit_day_changed<R: tauri::Runtime>(app: &tauri::AppHandle<R>) {
    if let Err(err) = app.emit("hm:day-changed", ()) {
        tracing::warn!(?err, "emit hm:day-changed failed");
    }
}
